src/Lib/Element.php

```diff
@@ -14,185 +14,207 @@
-        }
+        return $this->externalOptions;
     }
 
     /**
-     * Add Elements to an array
-     *
-     * @param  array  $serializable
-     *
-     * @return array
+     * Returns whether this element contains options
+     * @return bool
      */
-    private function processElementsForArray(array $serializable)
+    public function hasOptions()
     {
-        foreach ($this->elements as $key => $element) {
-            $serializable[$key] = $element;
-        }
-        return $serializable;
+        return count($this->options) > 0;
     }
 
     /**
-     * Add Elements to an Element
-     *
-     * @param  Element $serializable
-     *
-     * @return Element
+     * Returns whether this element conatins elements
+     * @return bool
      */
-    private function processElementsForObject(Element $serializable)
+    public function hasElements()
     {
-        foreach ($this->elements as $key => $element) {
-            $serializable->addElement($key, $element);
-        }
-        return $serializable;
+        return count($this->elements) > 0;
     }
 
     /**
-     * Add Elements to a string
-     *
-     * @param  string $serializable
-     *
-     * @return mixed
+     * Returns whether this object has anything to serialise
+     * @return bool
      */
-    private function processElementsForString($serializable)
+    public function isEmpty()
     {
-        if (count($this->elements) === 0) {
-            return $serializable;
-        } else {
-            return [
-                $serializable => $this->elements,
-            ];
-        }
+        return !$this->hasOptions() && !$this->hasElements();
     }
 
+
     /**
-     * Determine the type of serializable we have been passed, and append options to it accordingly
-     *
-     * @param  mixed $serializable
-     *
-     * @throws Exception\NotSupported
-     *
-     * @return mixed
+     * Extract any external elements from the elements and combine them with the elements array
+     * @return array
      */
-    private function processOptions($serializable)
+    protected function extractElements()
     {
-        if (is_array($serializable)) {
-            return $this->processOptionsForArray($serializable);
-        } elseif (is_object($serializable)) {
-            return $this->processOptionsForObject($serializable);
-        } elseif (is_string($serializable)) {
-            return $this->processOptionsForString($serializable);
-        } else {
-            throw new Exception\NotSupported('Cannot process options for var of type "%s"', gettype($serializable));
+        $elements = [];
+
+        // Fetch external elements from the elements
+        foreach($this->elements as $element) {
+            $elements = array_merge($elements, $element->getExternalElements());
         }
+
+        //Merge them with our elements
+        $elements = array_merge($this->elements, $elements);
```

Comment:
Not sure I understand this... you're foreaching ```$this->elements``` and merging on to ```$elements``` before, then you merge again at the top level?